src/non_nil.rs

```diff
@@ -0,0 +1,84 @@
+//! A wrapper type for nil UUIDs that provides a more memory-efficient
+//! `Option<NonNilUuid>` representation.
+
+use std::{fmt, num::NonZeroU128};
+
+use crate::Uuid;
+
+/// A UUID that is guaranteed not to be the nil UUID.
+///
+/// This is useful for representing optional UUIDs more efficiently, as `Option<NonNilUuid>`
+/// takes up the same space as `Uuid`.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct NonNilUuid(NonZeroU128);
+
+impl fmt::Display for NonNilUuid {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", Uuid::from(*self))
+    }
+}
+
+impl From<NonNilUuid> for Uuid {
+    /// Converts a [`NonNilUuid`] back into a [`Uuid`].
+    ///
+    /// # Examples
+    /// ```
+    /// use uuid::{non_nil::NonNilUuid, Uuid};
+    ///
+    /// let uuid = Uuid::from_u128(0x0123456789abcdef0123456789abcdef);
+    /// let non_nil = NonNilUuid::from(uuid);
+    /// let uuid_again = Uuid::from(non_nil);
+    ///
+    /// assert_eq!(uuid, uuid_again);
+    /// ```
+    fn from(non_nil: NonNilUuid) -> Self {
+        Uuid::from_u128(non_nil.0.get())
+    }
+}
+
+impl From<Uuid> for NonNilUuid {
```

Comment:
I don't think we should have a panicking `From` conversion and should instead use `TryFrom` like I think you had originally.

Reply:
fixed